### ISSUE TITLE
Optimization of scalars: Have `VariadicFunc` versions of associative/commutative binary functions

### ISSUE THREAD

**Issue by wangandi** [MaterializeInc/materialize#9345]
Currently, associative/commutative operators like `OR/AND/ADD/MUL` are binary (that is, variants of `BinaryFunc`), which means that we will have many layers of nesting if there are multiple clauses, addends, etc.

For example, the clause `substring(c_phone, 1, 2) IN ('13', '31', '23', '29', '30', '18', '17')` in TPCH 22 currently shows up in our plans as
```
(((((((#8 = "13") || (#8 = "31")) || (#8 = "23")) || (#8 = "29")) || (#8 = "30")) || (#8 = "18")) || (#8 = "17"))
```
(Note that the fact we rewrite IN into a chain of ORs is a separate issue).

Having `OR/AND/ADD/MUL` as a VariadicFunc has the following benefits:
* avoid stack overflow.
* allows for a more pleasing presentation of plans containing multiple clauses, addends, factors, etc.
* may make it easier to find common subexpressions.

**Comment by frankmcsherry** [comment]
I think partly addressed by https://github.com/MaterializeInc/materialize/pull/13731

**Comment by aalexandrov** [comment]
Correct. The rest will be addressed in #13931, so I'm going to close this one.

**Comment by frankmcsherry** [comment]
Wait, #13731 has already merged and hasn't done the rest.
